enhancements/authentication/allow-users-to-manage-their-own-tokens.md

```diff
@@ -0,0 +1,202 @@
+
+### Non-Goals
+
+- Changing the granularity of tokens to sessions: tokens can be used by many sessions.
+  The oauth system is not aware of sessions and won't be with this enhancement.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+A user is able to list/watch and delete (in both CLI and the web console) all of
+the tokens that were issued on their behalf.
+
+#### Story 2
+The user can look for specific details of their token like the name of the OAuth2
+client that requested issuing the token, token scopes, token expiration date etc.,
+making it easier for them to  audit their accesses.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The OAuth API server gets a new API endpoint - `useroauthaccesstokens`. This endpoint
+allows to `get`, `list`, `watch` and `delete` tokens that belong to the current user.
+The user is determined based on the context of the request, as supplied by the generic
+API server logic. This keeps impersonation working, so a user impersonating another
+one is capable to perform the above actions on behalf of the impersonated user.
+
+Upon receiving a REST request to the new endpoint, the API server internally
+wires the get/list/watch/delete requests to the `oauthaccesstoken` API endpoint.
+It then converts the results and returns them as `useroauthaccesstoken` type.
+The schema of the `useroauthaccesstoken` type is the same as of `oauthaccesstoken`,
+they only differ in kind.
+
+To list/watch the tokens internally, a field selector of the form `userName=<userName>`
+is used. It is wrapped in an and-type of a selector around any field selector
+user specifies.
+
+When getting or deleting a token, there must be an exact match of `token.userName`
+and the `userName` retrieved as described above. If such a match does not occur,
+the API server must return "404 - Not Found" so that it is not possible to guess
+other people's token object names.
+
+#### Roles modifications
+
+The `basic-user` clusterrole should now additionally receive the following rule:
+```
+apiGroups:
+- oauth.openshift.io
+resources:
+- useroauthaccesstokens
+verbs:
+- get
+- list
+- watch
+- delete
+```
+
+Since the new API allows to remove user-owned access tokens, it should no longer
+be necessary to keep the `system:oauth-token-deleter` clusterrole bound to the
+`system:authenticated` group. On the contrary, the currently existing clusterrole
+binding means a remaining, small security hole where a user can delete tokens
+that he only knows the sha256 hash of (but it cannot be used to login). To address
+this security hole, the clusterrolebinding `system:oauth-token-deleters` will
+therefore get deprecated as described in
+[Graduation Criteria - Removing a Deprecated Feature](#graduation-criteria---removing-a-deprecated-feature).
+
+#### Default table view in CLI
+
+By default, the following fields will be displayed when listing tokens in the
+command line:
+
+- metadata.name (`NAME`)
+- clientName (`CLIENT NAME`)
```

Comment:
@stlaz what will you see here in practice in a real cluster?

Reply:
@Anandnatraj this is the field I briefly mentioned. It will probably be a constant string like "console" or so, even if the token is used with oc.